turbopack/crates/turbopack-core/src/code_builder.rs

```diff
@@ -197,3 +236,68 @@ impl Code {
         Vc::cell(hash)
     }
 }
+
+#[turbo_tasks::function]
+pub async fn normalize_source_uris(
+    code: Vc<Code>,
+    context_path: Vc<FileSystemPath>,
+) -> Result<Vc<Code>> {
+    let mut builder = CodeBuilder::default();
+    let map = normalize_source_map(code.generate_source_map(), context_path);
+    let map = map.await?.map(Vc::upcast);
+    let code = &*code.await?;
+    builder.push_source(code.source_code(), map);
+    Ok(builder.build().cell())
+}
+
+#[turbo_tasks::function]
+pub async fn normalize_source_map(
+    map: Vc<OptionSourceMap>,
+    context_path: Vc<FileSystemPath>,
+) -> Result<Vc<OptionSourceMap>> {
+    let Some(map) = &*map.await? else {
+        return Ok(OptionSourceMap::none());
+    };
+
+    let flattened = map.await?.to_source_map().await?;
+    let flattened = flattened.as_regular_source_map();
+
+    let Some(flattened) = flattened else {
+        return Ok(OptionSourceMap::none());
+    };
+
+    let mut builder = SourceMapBuilder::new(flattened.get_file());
+    builder.set_debug_id(flattened.get_debug_id());
+    builder.set_source_root(flattened.get_source_root());
+    let context_fs = context_path.fs();
+    let context_fs = &*Vc::try_resolve_downcast_type::<DiskFileSystem>(context_fs)
+        .await?
+        .context("Expected the chunking context to have a DiskFileSystem")?
+        .await?;
+
+    let prefix = format!("{}[{}]/", SOURCE_MAP_PREFIX, context_fs.name);
+    for (src, src_contents) in flattened.sources().zip(flattened.source_contents()) {
+        let src = {
+            match src.strip_prefix(&prefix) {
+                Some(src) => {
+                    format!(
+                        "file://{}",
+                        sys_to_unix(
+                            &context_fs
+                                .to_sys_path(context_path.join(src.into()))
+                                .await?
+                                .to_string_lossy()
+                        )
```

Comment:
This need to be URI encoded to cover cases where e. g. there are spaces in filenames or `?` in the project path